In [1]:
import pandas as pd
import numpy as np

## Обзор данных

Для считывания данных используется функция `pd.read_csv`. У этой функции много аргументов (см. документацию), первый из которых - название считываемого файла с данными, второй - sep - разделитель, по умолчанию это запятая. Данные в csv файлах (обычно, но не всегда) записаны в виде "таблицы". Одна запись - одна строка, данные столбцов разделяются запятыми.

Для работы с файлом его, как и обычный текстовый файл (впрочем, это и есть обычный текстовый файл!), удобнее положить в ту же директорию, где лежит ноутбук. Затем вы можете прямо через jupyter открыть файл `csv` и увидеть, как он выглядит.

In [2]:
f = pd.read_csv("telecom_churn.csv")

Посмотреть начало и конец таблицы (по умолчанию 5 строк, но это настраиваемый параметр) - функции `head` и `tail`. Описание данных - `describe`, `info`. Размер таблицы - `shape`, названия колонок - `columns`. Обратите внимание, что `columns` возвращает итерируемый объект.

In [ ]:
f.head()

In [5]:
f.head(7)

State  Account length  Area code International plan Voice mail plan  \
0    KS             128        415                 No             Yes   
1    OH             107        415                 No             Yes   
2    NJ             137        415                 No              No   
3    OH              84        408                Yes              No   
4    OK              75        415                Yes              No   
5    AL             118        510                Yes              No   
6    MA             121        510                 No             Yes   

   Number vmail messages  Total day minutes  Total day calls  \
0                     25              265.1              110   
1                     26              161.6              123   
2                      0              243.4              114   
3                      0              299.4               71   
4                      0              166.7              113   
5                      0              223.4               98   
6                     24              218.2               88   

   Total day charge  Total eve minutes  Total eve calls  Total eve charge  \
0             45.07              197.4               99             16.78   
1             27.47              195.5              103             16.62   
2             41.38              121.2              110             10.30   
3             50.90               61.9               88              5.26   
4             28.34              148.3              122             12.61   
5             37.98              220.6              101             18.75   
6             37.09              348.5              108             29.62   

   Total night minutes  Total night calls  Total night charge  \
0                244.7                 91               11.01   
1                254.4                103               11.45   
2                162.6                104                7.32   
3                196.9                 89                8.86   
4                186.9                121                8.41   
5                203.9                118                9.18   
6                212.6                118                9.57   

   Total intl minutes  Total intl calls  Total intl charge  \
0                10.0                 3               2.70   
1                13.7                 3               3.70   
2                12.2                 5               3.29   
3                 6.6                 7               1.78   
4                10.1                 3               2.73   
5                 6.3                 6               1.70   
6                 7.5                 7               2.03   

   Customer service calls  Churn  
0                       1  False  
1                       1  False  
2                       0  False  
3                       2  False  
4                       3  False  
5                       0  False  
6                       3  False

In [ ]:
f.tail()

In [6]:
f.tail(7)

State  Account length  Area code International plan Voice mail plan  \
3326    OH              96        415                 No              No   
3327    SC              79        415                 No              No   
3328    AZ             192        415                 No             Yes   
3329    WV              68        415                 No              No   
3330    RI              28        510                 No              No   
3331    CT             184        510                Yes              No   
3332    TN              74        415                 No             Yes   

      Number vmail messages  Total day minutes  Total day calls  \
3326                      0              106.6              128   
3327                      0              134.7               98   
3328                     36              156.2               77   
3329                      0              231.1               57   
3330                      0              180.8              109   
3331                      0              213.8              105   
3332                     25              234.4              113   

      Total day charge  Total eve minutes  Total eve calls  Total eve charge  \
3326             18.12              284.8               87             24.21   
3327             22.90              189.7               68             16.12   
3328             26.55              215.5              126             18.32   
3329             39.29              153.4               55             13.04   
3330             30.74              288.8               58             24.55   
3331             36.35              159.6               84             13.57   
3332             39.85              265.9               82             22.60   

      Total night minutes  Total night calls  Total night charge  \
3326                178.9                 92                8.05   
3327                221.4                128                9.96   
3328                279.1                 83               12.56   
3329                191.3                123                8.61   
3330                191.9                 91                8.64   
3331                139.2                137                6.26   
3332                241.4                 77               10.86   

      Total intl minutes  Total intl calls  Total intl charge  \
3326                14.9                 7               4.02   
3327                11.8                 5               3.19   
3328                 9.9                 6               2.67   
3329                 9.6                 4               2.59   
3330                14.1                 6               3.81   
3331                 5.0                10               1.35   
3332                13.7                 4               3.70   

      Customer service calls  Churn  
3326                       1  False  
3327                       2  False  
3328                       2  False  
3329                       3  False  
3330                       2  False  
3331                       2  False  
3332                       0  False

In [7]:
f.tail(100)

State  Account length  Area code International plan Voice mail plan  \
3233    OK             112        415                 No              No   
3234    DE              75        510                 No             Yes   
3235    AZ              97        408                 No             Yes   
3236    AK             121        408                 No             Yes   
3237    MI             142        415                Yes              No   
...    ...             ...        ...                ...             ...   
3328    AZ             192        415                 No             Yes   
3329    WV              68        415                 No              No   
3330    RI              28        510                 No              No   
3331    CT             184        510                Yes              No   
3332    TN              74        415                 No             Yes   

      Number vmail messages  Total day minutes  Total day calls  \
3233                      0              166.0               79   
3234                     28              200.6               96   
3235                     25              141.0              101   
3236                     34              245.0               95   
3237                      0              140.8              140   
...                     ...                ...              ...   
3328                     36              156.2               77   
3329                      0              231.1               57   
3330                      0              180.8              109   
3331                      0              213.8              105   
3332                     25              234.4              113   

      Total day charge  Total eve minutes  Total eve calls  Total eve charge  \
3233             28.22               74.6              100              6.34   
3234             34.10              164.1              111             13.95   
3235             23.97              212.0               85             18.02   
3236             41.65              216.9               66             18.44   
3237             23.94              228.6              119             19.43   
...                ...                ...              ...               ...   
3328             26.55              215.5              126             18.32   
3329             39.29              153.4               55             13.04   
3330             30.74              288.8               58             24.55   
3331             36.35              159.6               84             13.57   
3332             39.85              265.9               82             22.60   

      Total night minutes  Total night calls  Total night charge  \
3233                247.9                 74               11.16   
3234                169.6                153                7.63   
3235                175.2                138                7.88   
3236                112.4                125                5.06   
3237                152.9                 88                6.88   
...                   ...                ...                 ...   
3328                279.1                 83               12.56   
3329                191.3                123                8.61   
3330                191.9                 91                8.64   
3331                139.2                137                6.26   
3332                241.4                 77               10.86   

      Total intl minutes  Total intl calls  Total intl charge  \
3233                 6.3                 7               1.70   
3234                 2.5                 5               0.68   
3235                 4.9                 2               1.32   
3236                 7.5                 8               2.03   
3237                10.9                 7               2.94   
...                  ...               ...                ...   
3328                 9.9                 6            

In [8]:
f.shape

(3333, 20)

In [9]:
f.columns

Index(['State', 'Account length', 'Area code', 'International plan',
       'Voice mail plan', 'Number vmail messages', 'Total day minutes',
       'Total day calls', 'Total day charge', 'Total eve minutes',
       'Total eve calls', 'Total eve charge', 'Total night minutes',
       'Total night calls', 'Total night charge', 'Total intl minutes',
       'Total intl calls', 'Total intl charge', 'Customer service calls',
       'Churn'],
      dtype='object')

In [10]:
f.describe()

Account length    Area code  Number vmail messages  Total day minutes  \
count     3333.000000  3333.000000            3333.000000        3333.000000   
mean       101.064806   437.182418               8.099010         179.775098   
std         39.822106    42.371290              13.688365          54.467389   
min          1.000000   408.000000               0.000000           0.000000   
25%         74.000000   408.000000               0.000000         143.700000   
50%        101.000000   415.000000               0.000000         179.400000   
75%        127.000000   510.000000              20.000000         216.400000   
max        243.000000   510.000000              51.000000         350.800000   

       Total day calls  Total day charge  Total eve minutes  Total eve calls  \
count      3333.000000       3333.000000        3333.000000      3333.000000   
mean        100.435644         30.562307         200.980348       100.114311   
std          20.069084          9.259435          50.713844        19.922625   
min           0.000000          0.000000           0.000000         0.000000   
25%          87.000000         24.430000         166.600000        87.000000   
50%         101.000000         30.500000         201.400000       100.000000   
75%         114.000000         36.790000         235.300000       114.000000   
max         165.000000         59.640000         363.700000       170.000000   

       Total eve charge  Total night minutes  Total night calls  \
count       3333.000000          3333.000000        3333.000000   
mean          17.083540           200.872037         100.107711   
std            4.310668            50.573847          19.568609   
min            0.000000            23.200000          33.000000   
25%           14.160000           167.000000          87.000000   
50%           17.120000           201.200000         100.000000   
75%           20.000000           235.300000         113.000000   
max           30.910000           395.000000         175.000000   

       Total night charge  Total intl minutes  Total intl calls  \
count         3333.000000         3333.000000       3333.000000   
mean             9.039325           10.237294          4.479448   
std              2.275873            2.791840          2.461214   
min              1.040000            0.000000          0.000000   
25%              7.520000            8.500000          3.000000   
50%              9.050000           10.300000          4.000000   
75%             10.590000           12.100000          6.000000   
max             17.770000           20.000000         20.000000   

       Total intl charge  Customer service calls  
count        3333.000000             3333.000000  
mean            2.764581                1.562856  
std             0.753773                1.315491  
min             0.000000                0.000000  
25%             2.300000                1.000000  
50%             2.780000                1.000000  
75%             3.270000                2.000000  
max             5.400000                9.000000

In [11]:
f.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3333 entries, 0 to 3332
Data columns (total 20 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   State                   3333 non-null   object 
 1   Account length          3333 non-null   int64  
 2   Area code               3333 non-null   int64  
 3   International plan      3333 non-null   object 
 4   Voice mail plan         3333 non-null   object 
 5   Number vmail messages   3333 non-null   int64  
 6   Total day minutes       3333 non-null   float64
 7   Total day calls         3333 non-null   int64  
 8   Total day charge        3333 non-null   float64
 9   Total eve minutes       3333 non-null   float64
 10  Total eve calls         3333 non-null   int64  
 11  Total eve charge        3333 non-null   float64
 12  Total night minutes     3333 non-null   float64
 13  Total night calls       3333 non-null   int64  
 14  Total night charge      3333 non-null   

Небольшое, но важное отступление. Глубоко понимать это пока не обязательно. Обратите внимание, что функция `print` возвращает строковое представление объекта и она не необходима, чтобы вывести объект. 2 ячейки сработают одинаково:

In [4]:
a = 1
print(a)

1


In [3]:
a = 1
a

1

При этом для некоторых объектов (для которых стандартные функции `__str__` и `__repr__` возвращают разные значения) два способа вывода сработают несколько по-разному:

In [12]:
a = np.array([1.0, 2.0])
print(a)

[1. 2.]


In [13]:
a = np.array([1.0, 2.0])
a

array([1., 2.])

Такая же ситуация с pandas dataframe'ами. Без print получится красивее, чем с ним, но фактически это одно и то же:

In [14]:
print(f)

     State  Account length  Area code International plan Voice mail plan  \
0       KS             128        415                 No             Yes   
1       OH             107        415                 No             Yes   
2       NJ             137        415                 No              No   
3       OH              84        408                Yes              No   
4       OK              75        415                Yes              No   
...    ...             ...        ...                ...             ...   
3328    AZ             192        415                 No             Yes   
3329    WV              68        415                 No              No   
3330    RI              28        510                 No              No   
3331    CT             184        510                Yes              No   
3332    TN              74        415                 No             Yes   

      Number vmail messages  Total day minutes  Total day calls  \
0                   

In [15]:
f

State  Account length  Area code International plan Voice mail plan  \
0       KS             128        415                 No             Yes   
1       OH             107        415                 No             Yes   
2       NJ             137        415                 No              No   
3       OH              84        408                Yes              No   
4       OK              75        415                Yes              No   
...    ...             ...        ...                ...             ...   
3328    AZ             192        415                 No             Yes   
3329    WV              68        415                 No              No   
3330    RI              28        510                 No              No   
3331    CT             184        510                Yes              No   
3332    TN              74        415                 No             Yes   

      Number vmail messages  Total day minutes  Total day calls  \
0                        25              265.1              110   
1                        26              161.6              123   
2                         0              243.4              114   
3                         0              299.4               71   
4                         0              166.7              113   
...                     ...                ...              ...   
3328                     36              156.2               77   
3329                      0              231.1               57   
3330                      0              180.8              109   
3331                      0              213.8              105   
3332                     25              234.4              113   

      Total day charge  Total eve minutes  Total eve calls  Total eve charge  \
0                45.07              197.4               99             16.78   
1                27.47              195.5              103             16.62   
2                41.38              121.2              110             10.30   
3                50.90               61.9               88              5.26   
4                28.34              148.3              122             12.61   
...                ...                ...              ...               ...   
3328             26.55              215.5              126             18.32   
3329             39.29              153.4               55             13.04   
3330             30.74              288.8               58             24.55   
3331             36.35              159.6               84             13.57   
3332             39.85              265.9               82             22.60   

      Total night minutes  Total night calls  Total night charge  \
0                   244.7                 91               11.01   
1                   254.4                103               11.45   
2                   162.6                104                7.32   
3                   196.9                 89                8.86   
4                   186.9                121                8.41   
...                   ...                ...                 ...   
3328                279.1                 83               12.56   
3329                191.3                123                8.61   
3330                191.9                 91                8.64   
3331                139.2                137                6.26   
3332                241.4                 77               10.86   

      Total intl minutes  Total intl calls  Total intl charge  \
0                   10.0                 3               2.70   
1                   13.7                 3               3.70   
2                   12.2                 5               3.29   
3                    6.6                 7               1.78   
4                   10.1                 3               2.73   
...                  ...               ...                ...   
3328                 9.9                 6            

2 варианта синтаксиса для просмотра содержимого столбца:

In [ ]:
f.State

In [ ]:
f['State']

Обратите внимание, что таблица и столбцы имеют разный тип:

In [ ]:
type(f), type(f.State), type(f['State'])

Тем не менее, мы можем сделать именно датафрейм из одного столбца следующим образом:

In [ ]:
f[['State']]

In [ ]:
type(f[['State']]), type(f['State'])

Это способ обращения к "срезам" таблицы по названиям столбцов. Еще пример:

In [ ]:
f[['State', 'Churn']]

Просмотр уникальных значений в столбце и их числа:

In [ ]:
f.State.unique()

In [ ]:
f.State.nunique()

Обращение к элементу ообъекта `pd.Series`:

In [ ]:
f.State[0]

Полный набор функций, применимых к столбцу можно посмотреть следующим образом. Здесь используется стандартная функция `__dir__()`, которая выводит список атрибутов и методов объекта.

In [ ]:
f.State.__dir__()

С `pd.DataFrame` привычный способ обращения по индексам не сработает:

In [ ]:
f[0][0]

Основной способ - использование `loc` и `iloc`. `loc` дает индексацию по "настоящим" индексам - названиям столбцов и названиям строк. В нашем случае, названия строк - целые числа, как обычные индексы. `iloc` позволяет осуществлять стандартную численную индексацию, как для `numpy` массивов.

In [ ]:
f.iloc[0] # первая строка

In [ ]:
f.iloc[0][0]

In [ ]:
f.iloc[0, 0]

In [ ]:
f.iloc[[0, 0]]

In [ ]:
f.iloc[[1, 2, 5, 1]]

In [ ]:
f.loc[[1, 2, 5, 1]]

In [ ]:
f.loc[[1, 2, 5, 1], ["State", "Account length"]]

In [ ]:
f.iloc[[1, 2, 5, 1], [0, 1]]

Следующая ячейка вернет ошибку:

In [ ]:
f.iloc[[1, 2, 5, 1], ["State", "Account length"]]

In [ ]:
f.iloc[:5, :4]

In [ ]:
f.loc[:, ["State", "Account length"]]

In [ ]:
f.loc[4:9, ["State", "Account length", "Area code", "International plan"]]

Обратите внимание, что `loc` включает и строки соответствующие и начальному, и конечному индексам.

Сохраним этот датафрейм в новую переменную и рассмотрим отличие loc и iloc в работе по целочисленным индексам строк

In [ ]:
new_f = f.loc[4:9, ["State", "Account length", "Area code", "International plan"]]

In [ ]:
new_f.iloc[:5]

In [ ]:
new_f.loc[:5]

**Упражнение 1.** Объясните различие поведения `loc` и `iloc` для `new_f`.

#### Создание датафрейма и присваивание по индексам

In [ ]:
df = pd.DataFrame({"A": [1, 2, 3], "B": [4, 5, 6]}, index = ["first", "second", "third"])
df

In [ ]:
df.iloc[0, 0] = 100
df

In [ ]:
df.loc["second", "B"] = 31
df

In [ ]:
df.iloc[2] = 5
df

In [ ]:
df.loc["B"] = 0
df

In [ ]:
df.loc[:, "B"] = 0
df

In [ ]:
df = pd.DataFrame({"A": [1, 2, 3, 4], "B": [5, 6, 7, 8]})
df

In [ ]:
df.iloc[1:2, :] = -1
df

In [ ]:
df.loc[1:2, :] = -2
df

Еще раз убеждаемся, что `loc` включает оба конца "среза", а `iloc` - полуинтервал, как например обычные срезы.

#### Создание нового столбца

In [ ]:
df["C"] = 1
df

In [ ]:
df["D"] = df["A"] + df["B"]
df

In [ ]:
df["E"] = df["D"].apply(lambda x: x ** 2)
df

In [ ]:
df["F"] = df[["A", "C", "E"]].apply(lambda x: x.sum(), axis=1)
df

**Упражнение 2.** Объясните последний результат.

## Срезы с условиями и аггрегация

In [ ]:
f[f.Churn == True]

In [ ]:
f[f["Area code"] == 408]

In [ ]:
f[f["Area code"] == 408].loc[3]

In [ ]:
f[f["Area code"] == 408].loc[3] == f[f["Area code"] == 408].iloc[0]

In [ ]:
f[f.Churn == True]

In [ ]:
f[f.Churn == True].mean()

In [ ]:
f[f.Churn == True].sum()

Выборка по нескольким условиям требует скобок, т.к. операции `&`, `|`, `~` (и, или, не) имеют более высокий приоритет, чем `==`, `>=`, `<=` и арифметические:

In [ ]:
f[(f.Churn == True) & (f["Area code"] == 408)]

Можем посмотреть средние показатели лояльных и нелояльных клиентов (столбец "Churn" показывает, ушел клиент от нас как от оператора связи или нет).

In [ ]:
loyal = f[f.Churn == False].mean()
unloyal = f[f.Churn == True].mean()
pd.DataFrame({"loyal": loyal, "unloyal": unloyal})

`groupby` позволяет делать группировку по уникальным значениям столбца, которые будут играть роль индекса в полученном датафрейме. При этом также нужна функция агрегации (`mean`, `sum`, `max` и прочие), чтобы отобразить полученный объект.

In [ ]:
f.groupby("Churn")

In [ ]:
f.groupby("Churn").mean()

Здесь мы сделали агрегацию всех столбцов функцией `mean`. Часто нам нужна агрегация по конкретному или конкретным столбцам. Посмотрим, например, в каком штате больше всего разговаривают по телефону:

In [ ]:
f.groupby(["State"]).agg({"Total day minutes": "mean"})

Максимальное значение времени:

In [ ]:
f.groupby(["State"]).agg({"Total day minutes": "mean"}).max()

Чтобы найти сам штат, можем отсортировать данные:

In [ ]:
f.groupby(["State"]).agg({"Total day minutes": "mean"}).sort_values(by="Total day minutes", ascending = False)

**Пример.** Как распределен отток клиентов по штатам? В каких штатах он больше среднего?

In [ ]:
f.groupby(["State"]).agg({"Churn": "mean"})

In [ ]:
f.groupby(["State"]).agg({"Churn": "mean"}).loc["LA"]

In [ ]:
f.groupby(["State"]).agg({"Churn": "mean"}).mean()

In [ ]:
f.groupby(["State"]).agg({"Churn": "mean"}).mean()[0]

In [ ]:
m = f.groupby(["State"]).agg({"Churn": "mean"}).mean()[0]
new_dataframe = f.groupby(["State"]).agg({"Churn": "mean"})
new_dataframe[new_dataframe.Churn > m]

## Упражнения.

**Упражнение 3.** Найдите среднее количество звонков `Total day calls` для всего датафрейма.

**Упражнение 4.** Найдите среднее количество звонков `Total day calls` для любого выбранного вами штата.

**Упражнение 5.** Создайте датафрейм, в котором будет среднее количество звонков `Total day calls` для каждого штата.

**Упражнение 6.** Оставьте в созданном датафрейме строки только с теми штатами, где количество звонков `Total day calls` больше среднего по исходному датафрейму.

**Упражнение 7.** Создайте датафрейм, в котором будует средние количества звонков `Total day calls` и `Total eve calls` для каждого штата.

**Упражнение 8.** Создайте датафрейм, в котором будует средние количества звонков `Total day calls` и `Total eve calls` для каждого штата, а также столбец со значениями True и False - ответом на вопрос, больше ли дневных звонков, чем вечерних.

**Упражнение 9.** Найти долю клиентов (отношение их числа к общему количеству клиентов) с `international plan` и `voice mail plan`.

**Упражнение 10.** Найти число уникальных значений `Area code`.

**Упражнение 11.** Вывести DataFrame из 2 столбцов: число звонков в поддержку; число клиентов, звонивших столько раз. Подсказка: используйте функцию агрегации `count`.

**Упражнение 12.** Вывести DataFrame из 2 столбцов: число звонков в поддержку; доля оттока (`Churn`). Построить график.

**Упражнение 13.** Найти среднюю длительность международного (`intl`) звонка.

**Упражнение 14*.** Какие звонки дольше - дневные, вечерние или ночные? Ответ привести в формате DataFrame $3*3$: строки - число минут, число звонков, среднее время звонка. Столбцы - день, вечер, ночь

**Упражнение 15.** Сравнить `Total day charge` для оставшихся и ушедших клиентов.

**Упражнение 16.** Отсортриуйте штаты по `Total day charge` (по возрастанию).

**Упражнение 17.** Сделайте агрегацию по средним показателям для каждой `Area code`.

**Упражнение 18.** Выведите датафрейм размера $3*2$: столбцы State, Churn для 100, 102 и 104 строк исходного датафрейма.

**Упражнение 19.** Создайте датафрейм из 2 столбцов, заполненных произвольными числами. Добавьте третий столбец, равный их сумме квадратов.

**Упражнение 20.** Добавьте к созданному датафрейму четвертый столбец, равный среднему значению первых трех, *используя* функцию `mean`.